src/jit/lower.cpp

```diff
@@ -2100,6 +2057,80 @@ void Lowering::LowerFastTailCall(GenTreeCall* call)
+            {
+                LclVarDsc* callerArgDsc = comp->lvaGetDesc(callerArgLclNum);
+
+                if (callerArgDsc->lvIsRegArg)
+                    continue;
+
+#if defined(_TARGET_WINDOWS_) && defined(_TARGET_64BIT_)
+                // On Win64, the argument position determines the stack slot uniquely, and even the
+                // register args take up space in the stack frame (shadow space).
+                unsigned int argStart = callerArgLclNum * TARGET_POINTER_SIZE;
+                unsigned int argEnd   = argStart + static_cast<unsigned int>(callerArgDsc->lvArgStackSize());
+#else
+                assert(callerArgDsc->lvStkOffs != BAD_STK_OFFS);
+
+                if (baseOff == -1)
+                    baseOff = callerArgDsc->lvStkOffs;
+
+                // On all ABIs where we fast tail call the stack args should come in order.
+                assert(baseOff <= callerArgDsc->lvStkOffs);
+
+                // Compute offset of this stack argument relative to the first stack arg.
+                // This will be its offset into the incoming arg space area.
+                unsigned int argStart = static_cast<unsigned int>(callerArgDsc->lvStkOffs - baseOff);
+                unsigned int argEnd   = argStart + comp->lvaLclSize(callerArgLclNum);
+#endif
+
+                // If ranges do not overlap then this PUTARG_STK will not mess up the arg.
+                if ((overwrittenEnd <= argStart) || (overwrittenStart >= argEnd))
+                    continue;
+
+                // Codegen cannot handle a disjoint overlapping copy. For example, if
+                // we have
+                // bar(S16 stack, S32 stack2)
+                // foo(S32 stack, S32 stack2) { bar(..., stack) }
+                // then we may end up having to move 'stack' in foo 16 bytes ahead.
+                // It is possible that this PUTARG_STK is the only use, in which case we will
+                // need to introduce a temp
+                // look for uses starting from it. Note that we assume non-disjoint copies are OK.
+                GenTree* lookForUsesFrom = put->gtNext;
+                if (overwrittenStart != argStart)
+                    lookForUsesFrom = insertionPoint;
+
+                RehomeArgForFastTailCall(callerArgLclNum, insertionPoint, lookForUsesFrom, call);
+                // The call can introduce temps and invalidate the pointer.
+                callerArgDsc = comp->lvaGetDesc(callerArgLclNum);
+
+                // For promoted locals we have more work to do as its fields could also have been invalidated.
+                if (!callerArgDsc->lvPromoted)
+                    continue;
+
+                unsigned int fieldsFirst = callerArgDsc->lvFieldLclStart;
+                unsigned int fieldsEnd   = fieldsFirst + callerArgDsc->lvFieldCnt;
+                for (unsigned int j = fieldsFirst; j < fieldsEnd; j++)
+                    RehomeArgForFastTailCall(j, insertionPoint, lookForUsesFrom, call);
```

Comment:
I am not 100% sure whether this is necessary -- before my change we were introducing temps for the first promoted field so I have left this behavior (in a slightly more robust fashion). I don't know if this should only be done in the dependent promotion case -- is this the only case where the field can live in the same slot as the parent struct?